File: discrete-math/lab06/H.cpp
#include <iostream>
#include <vector>
#include <set>

int n;

bool is_zero(const std::vector<std::set<int>>& graph) {
    for (int i = 0; i < n; ++i) {
        if (!graph[i].empty()) {
            return false;
        }
    }

    return true;
}

std::vector<int> chromatic_polynomial(const std::vector<std::set<int>>& graph, int countV) {
    std::vector<int> returnArray;
    returnArray.resize(n + 1);

    if (is_zero(graph)) {
        returnArray[countV] = 1;
        return returnArray;
    }

    std::vector<std::set<int>> graph1(graph), graph2(graph);

    int v = 0, u;

    while (graph[v].empty()) {
        v++;
    }

    u = *graph[v].begin();

    graph1[v].erase(u);
    graph1[u].erase(v);

    graph2[v].erase(u);
    graph2[u].erase(v);

    for (auto w : graph2[u]) {
        graph2[v].insert(w);
        graph2[w].insert(v);
        graph2[w].erase(u);
    }

    graph2[u].clear();

    std::vector<int> result1 = chromatic_polynomial(graph1, countV);
    std::vector<int> result2 = chromatic_polynomial(graph2, countV - 1);

    for (int i = 0; i <= n; i++) {
        returnArray[i] = result1[i] - result2[i];
    }

    return returnArray;
}

int main() {
    int m;
    std::cin >> n >> m;

    std::vector<std::set<int>> graph;
    graph.resize(n);

    for (int i = 0; i < m; i++) {
        int v, u;
        std::cin >> v >> u;
        v--;
        u--;

        graph[v].insert(u);
        graph[u].insert(v);
    }

    std::vector<int> result = chromatic_polynomial(graph, n);

    int index = n;
    while (result[index] == 0) {
        index--;
    }

    std::cout << index << '\n';
    for (int i = n; i >= 0; i--) {
        std::cout << result[i] << " ";
    }

    return 0;
}

File: discrete-math/lab06/H.kt
private fun readLn() = readLine()!! // string line
private fun readInt() = readLn().toInt() // single int
private fun readStrings() = readLn().split(" ") // list of strings
private fun readInts() = readStrings().map { it.toInt() } // list of ints

var n = 0

// Правильно, но медленно. Переписал на плюсы
fun main() {
    val nm = readInts()
    n = nm.first()
    val m = nm.last()
    val graph = Array(n) { HashSet<Int>() }

    repeat(m) {
        val vu = readInts()
        val v = vu.first() - 1
        val u = vu.last() - 1

        graph[v].add(u)
        graph[u].add(v)
    }

    val result: Array<Int> = chromaticPolynomial(graph, n)

    println(result.indexOfLast { it > 0 })
    println(result.reversed().joinToString(separator = " "))
}

fun chromaticPolynomial(graph2: Array<HashSet<Int>>, countV: Int): Array<Int> {
    val returnArray = Array(n + 1) { 0 }
    var zero = true

    for (i in 0 until n) {
        if (graph2[i].isNotEmpty()) {
            zero = false
            break
        }
    }

    if (zero) {
        returnArray[countV] = 1
        return returnArray
    }

    val graph1 = Array(n) { HashSet<Int>() }

    for (i in 0 until n) {
        graph1[i].addAll(graph2[i])
    }

    val v = graph1.indexOfFirst { it.size > 0 }
    val u = graph1[v].first()
    graph1[v].remove(u)
    graph1[u].remove(v)

    graph2[v].remove(u)
    graph2[u].remove(v)

    for (w in graph2[u]) {
        graph2[v].add(w)
        graph2[w].add(v)
        graph2[w].remove(u)
    }
    graph2[u].clear()

    val result1 = chromaticPolynomial(graph1, countV)
    val result2 = chromaticPolynomial(graph2, countV - 1)

    for (i in 0..n) {
        returnArray[i] = result1[i] - result2[i]
    }

    return returnArray
}